File: app/src/main/java/com/qimian233/ztool/hook/modules/mobiledesktop/DisableNearbyShareAutoOffHook.java
package com.qimian233.ztool.hook.modules.mobiledesktop;

import com.qimian233.ztool.hook.base.BaseHookModule;

import io.github.libxposed.api.XposedModuleInterface;

import java.lang.reflect.Method;

/**
 * 测试 Hook — 禁用超级互联附近分享的 10 分钟自动关闭倒计时。
 * <p>
 * 机制：{@code ra.c.q()} (FileUnionSwitchManager.startCountDown) 
 * 在附近分享开启后发送延迟消息 (what=1, delay=600000ms=10min)，
 * handler {@code d8.a.m()} 收到消息后调用 
 * {@code c0.setNearbyShareStatus(false)} 自动关闭。
 * 此 Hook 将 startCountDown 替换为空操作，阻止倒计时启动。
 * </p>
 */
public class DisableNearbyShareAutoOffHook extends BaseHookModule {

    private static final String TARGET_PACKAGE = "com.motorola.mobiledesktop";
    // ra.c = FileUnionSwitchManager
    private static final String START_COUNTDOWN_CLASS = "ra.c";

    public DisableNearbyShareAutoOffHook() {}

    @Override
    public String getModuleName() {
        return "test_hook";
    }

    @Override
    public String[] getTargetPackages() {
        return new String[]{TARGET_PACKAGE};
    }

    @Override
    public void handleLoadPackage(XposedModuleInterface.PackageLoadedParam param) throws Throwable {
        ClassLoader classLoader = param.getDefaultClassLoader();

        try {
            Class<?> fileUnionSwitchManagerClass = classLoader.loadClass(START_COUNTDOWN_CLASS);

            // Hook q() — startCountDown: ()V
            for (Method method : fileUnionSwitchManagerClass.getDeclaredMethods()) {
                if (method.getName().equals("q")
                        && method.getParameterTypes().length == 0
                        && method.getReturnType() == void.class) {
                    this.xposed.hook(method).intercept(chain -> {
                        log("startCountDown() intercepted — auto-off timer prevented.");
                        return null;
                    });
                    log("Installed hook for FileUnionSwitchManager.startCountDown()");
                    return;
                }
            }
            logError("Could not find startCountDown() method in ra.c", null);
        } catch (ClassNotFoundException e) {
            logError("ra.c (FileUnionSwitchManager) not found", e);
        } catch (Throwable t) {
            logError("Failed to hook FileUnionSwitchManager.startCountDown()", t);
        }
    }
}

File: app/src/main/java/com/qimian233/ztool/hook/modules/mobiledesktop/DisableNearbyShareAutoOffHook.kt
package com.qimian233.ztool.hook.modules.mobiledesktop

import com.qimian233.ztool.data.keys.ScopeKeys
import com.qimian233.ztool.dexindex.base.DexIndexConstants
import com.qimian233.ztool.hook.base.AppHookModule
import com.qimian233.ztool.hook.base.DexIndexStore
import io.github.libxposed.api.XposedModuleInterface

/**
 * Test Hook — disables the 10-minute auto-off countdown of nearby sharing
 * in Super Interconnect (FileUnion).
 *
 * Mechanism: startCountDown of FileUnionSwitchManager sends a
 * delayed 10-minute message after nearby sharing is enabled; when the handler
 * receives it, it turns nearby sharing off. This Hook replaces startCountDown
 * with a no-op to prevent the countdown from starting.
 *
 * The target method is obfuscated; it is located by the DexIndexer via the
 * log string "startCountDown()", falling back to the class/method names
 * hardcoded for the current version.
 */
class DisableNearbyShareAutoOffHook : AppHookModule() {

    companion object {
        private val TARGET_PACKAGE = ScopeKeys.MOBILE_DESKTOP.packageName
        // Fallback: hardcoded class/method names of FileUnionSwitchManager in the current version
        private const val FALLBACK_CLASS = "com.motorola.motoaccount.sdk.se.c"
        private const val FALLBACK_METHOD = "b"
    }

    override fun getModuleName(): String = "disable_nearby_share_countdown"

    override fun getTargetPackages(): Array<String> = arrayOf(TARGET_PACKAGE)

    override fun handleLoadPackage(param: XposedModuleInterface.PackageLoadedParam) {
        val classLoader = param.defaultClassLoader

        val module = DexIndexStore.lookup(xposed, ScopeKeys.MOBILE_DESKTOP.packageName)
            ?.getAsJsonObject(DexIndexConstants.ModuleKeys.DISABLE_NEARBY_SHARE_COUNTDOWN)
        val targetClassName = module?.get(DexIndexConstants.Keys.TARGET_CLASS)
            ?.takeIf { !it.isJsonNull }?.asString ?: FALLBACK_CLASS
        val targetMethodName = module?.get(DexIndexConstants.Keys.TARGET_METHOD)
            ?.takeIf { !it.isJsonNull }?.asString ?: FALLBACK_METHOD

        try {
            val targetClass = classLoader.loadClass(targetClassName)

            val targetMethod = targetClass.declaredMethods.firstOrNull { method ->
                method.name == targetMethodName
                        && method.parameterTypes.isEmpty()
                        && method.returnType == Void.TYPE
            }

            if (targetMethod == null) {
                logger.error(
                    "Could not find startCountDown method ($targetMethodName) in $targetClassName",
                    null
                )
                return
            }

            hookWithId(targetMethod, "target") { 
                logger.debug("startCountDown() intercepted — auto-off timer prevented.")
                null
            }
            logger.info("Installed hook for FileUnionSwitchManager.$targetMethodName()")
        } catch (e: ClassNotFoundException) {
            logger.error("$targetClassName (FileUnionSwitchManager) not found", e)
        } catch (t: Throwable) {
            logger.error("Failed to hook FileUnionSwitchManager.startCountDown()", t)
        }
    }
}
